base/abn_metrics.go

```diff
@@ -0,0 +1,166 @@
+package base
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	abnapp "github.com/iter8-tools/iter8/abn/application"
+	pb "github.com/iter8-tools/iter8/abn/grpc"
+	k8sclient "github.com/iter8-tools/iter8/abn/k8sclient"
+	log "github.com/iter8-tools/iter8/base/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"helm.sh/helm/v3/pkg/cli"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// abnClientInterface is interface for calling gRPC services
+type abnClientInterface interface {
+	callGetApplicationJSON(appName string) (string, error)
+}
+
+// defaultABNClient is default implementation of interface that calls the service
+type defaultABNClient struct {
+	endpoint string
+}
+
+func (wc *defaultABNClient) callGetApplicationJSON(appName string) (string, error) {
+	// setup client
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+	conn, err := grpc.Dial(wc.endpoint, opts...)
+	if err != nil {
+		return "", err
+	}
+	c := pb.NewABNClient(conn)
+
+	// get application
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	s, err := c.GetApplicationData(
+		ctx,
+		&pb.ApplicationRequest{
+			Application: appName,
+		},
+	)
+	if err != nil {
+		return "", err
+	}
+	applicationJSON := s.GetApplicationJson()
+
+	return applicationJSON, nil
+}
+
+const (
+	// CollectABNMetrics is the name of the task to read A/B(/n) metric values
+	CollectABNMetrics = "abnmetrics"
+	// abnMetricProvider is prefix for abn metrics
+	abnMetricProvider = "abn"
+)
+
+// ABNMetricsInputs is the inputs for for the abnmetrics task
+type ABNMetricsInputs struct {
+	// Application is name of application to evaluate
+	Application string `json:"application" yaml:"application"`
+}
+
+// collectABNMetricsTask is task defintion for abnmetrics task
+type collectABNMetricsTask struct {
+	TaskMeta
+	With   ABNMetricsInputs `json:"with" yaml:"with"`
+	client abnClientInterface
+}
+
+// initializeDefaults sets default values for the task
+func (t *collectABNMetricsTask) initializeDefaults() {
+	k8sclient.Client = *k8sclient.NewKubeClient(cli.New())
```

Comment:
Why does this task need to initialize a k8sclient? For instance, the `collect_http` task which also writes metrics, doesn't appear to require the client initialization.

Reply:
no longer needed. good catch. removed.